### PR TITLE
Test cases for Fallbacks container

### PR DESCRIPTION
**Pull request by j-kuehn**
I added three test cases for the `Fallbacks` container:

- `ConnectStageInsideFallbacks` shows that adding a `Connect` stage to the `Fallbacks` container seems to be broken, as no solution can be computed for the task currently.
- `ComputeFirstSuccessfulStageOnly` suggests that solutions inside the `Fallbacks` container should not be computed after one solution has been found. I'm not sure how the container was originally intended to work, so this might not be a bug.
- `ActiveChildReset` reveals the failure of the container to reset the [`active_child_`](https://github.com/ros-planning/moveit_task_constructor/blob/master/core/include/moveit/task_constructor/container.h#L161) for every spawned solution from a generator. In this test case, the first `ForwardMockup` does not become the `active_child_` for the third spawned solution from the `GeneratorMockup`.

### CODE REVIEW

File: core/test/test_container.cpp
Comment (v4hn) [inline core/test/test_container.cpp:736]: Depending on the preferred semantics (whether fallback alternatives should be computed at all), the test is still not correct: If no alternatives should be considered you would have to check for *two* solutions where both should come from the first `ForwardMockup`.

Let's wait for Robert's response before modifying this.

**Review by v4hn — COMMENTED** [review]
@j-kuehn Thank you for the discussed tests! I would appreciate if you could provide a fix for `ActiveChildReset` as well. :cat2: 

@rhaschke I would appreciate your opinion on the second issue. I strictly prefer the behavior in the test (don't run alternatives if a solution succeeds) but if you prefer we can add an option to toggle behaviors instead.

File: core/test/test_container.cpp
Comment (rhaschke) [reply core/test/test_container.cpp]: I don't yet fully understand the purpose of this test. I would expect two solutions here, contributed by the _first_ child. However, this was not yet tested here. I pushed a corresponding fixup.

File: core/test/test_container.cpp
Comment (rhaschke) [inline core/test/test_container.cpp:690]: This fails for the following reason:
1. The `Generator` spawns a solution, which is correctly forwarded to the start pull interface of `Connect`.
2. However, because there was no state pushed into the end pull interface of `Connect`, `canCompute()` returns false
and the `Fallbacks` container forwards to the next child (which doesn't exist).

This can only be solved, if we drop the simple linear execution order and switch to topological execution order.

**Review by rhaschke — APPROVED** [review]
Thanks for providing these tests. See detailed comments below.
I suggest to disable/skip `ConnectStageInsideFallbacks` for now as it won't easily fix.
@v4hn: Do you already have concrete plans, how to implement a topological processing order?

File: core/test/test_container.cpp
Comment (rhaschke) [reply core/test/test_container.cpp]: There should be another test case, where the external generator is failing first before succeeding.
Currently, I would expect this test to fail (with a solution of the wrong, second child).

**Comment by rhaschke** [comment]
Ping @v4hn, @j-kuehn. Any thoughts on my feedback?

File: core/test/test_container.cpp
Comment (j-kuehn) [reply core/test/test_container.cpp]: The test suggested that after computing solutions for the second spawned solution from the generator, the `active_child_` would not be reset to the first child for computing the third solution. Before your two commits, this led to the first child being ignored for the third spawned generator solution.

As @v4hn pointed out, this test was not consistent with the proposed functionality in `ComputeFirstSuccessfulStageOnly`, but rather tried to test the status quo. But even after your fix for the `ActiveChildReset` test, I would have expected the test to fail, because as I understand it, the `active_child_` is pointing to the second child after the second spawned generator solution. So for the third solution, I expected the container to use the second child for computation. However, the test passes, so it is actually using the first child. I'm not sure how your commit fixed this.

I just tried adding another test as you suggested, where the external generator is failing first before succeeding:
```cpp
std::make_unique<GeneratorMockup>(PredefinedCosts{ { INF, 0.0 } }));
```
However, this yields 0 solutions and I'm not sure why.

File: core/test/test_container.cpp
Comment (j-kuehn) [reply core/test/test_container.cpp:690]: Sounds reasonable to disable this test for now.

**Review by v4hn — COMMENTED** [review]
Thanks for the feedback. I wanted to leave it to @j-kuehn to implement a fix for at least the simple test, but it looks like we disagree on what exactly should happen? Please give your reasoning to my comment below.

**Comment by v4hn** [comment]
superseded by #287

**Pull request closed without merging**
